## CNView visualize wrapper

Note: To convert this into an R file, use `knitr::purl("CNView_wrapper.Rmd", output="CNView_wrapper.R", documentation = 2)`

Check for necessary packages and install as needed
```{r}
##Loads required packages; installs if necessary##
cran.package.list <- c("readxl","pdftools","RMySQL", "data.table",
                       "plyr","MASS","metap")
for(i in cran.package.list){
   if(!(i %in% rownames(installed.packages()))){
     message('Installing package: ',i)
     install.packages(i,dependencies = T)
   } else next
}

library(pdftools)
library(readxl)
library(metap)
##Loads required packages##
library(RMySQL)
library(plyr)
library(MASS)
#require(RMySQL)
#require(plyr)
#require(MASS)
#require(Rsamtools)

```

## SPECIFY SOURCE CNVIEW PROGRAM
```{r}
#source('CNVIew_FIXED_SS_04_15_2018_main.R')
#source('CNView_srv05_orig.R')
source('CNViewMod.R')
```

## SPECIFY INPUT AND OUTPUT PARAMETERS
Modify here as needed. Any file can be selected interactively if `file.choose()` function is used, e.g.:
`regions_file <- file.choose()`

__Note:__ chromosomes must NOT have the 'chr' prefix in both the `regions_file` and `coverage_matrix` input data

```{r}
##-- Excel file with CNV regions to view (must have columns: chrom, start, stop)
regions_file <- 'INPUT_test_18regions.xlsx'
#regions_file <- file.choose()

##-- SPECIFY BRAINS TO VIEW
common_brains=c("LIBD61_NeuN_positive", "LIBD62_NeuN_negative")
#common_brains=c("LIBD08_Cerebellum", "LIBD25_Cerebellum")

##-- SPECIFY COVERAGE MATRIX
coverage_matrix='UO1_merged_CNView_input.txt'
#coverage_matrix <- file.choose()
#coverage_matrix='UO1_30x_CNView_input.txt'

##-- SPECIFY output directory (can be a full path)
outdir = 'out_UO1_mrg'
# create this directory if it doesn't exist:
if (!dir.exists(outdir)){
 dir.create(outdir)
}

##-- SPECIFY output file suffix
out_suffix <- "_CNView_18regions_UO1_merged"
## the final format is going to be: {outdir}/chr_start_stop_{out_suffix}
```

## Load and process data
```{r}
## load the Excel file:
CNVs = as.data.frame(read_xlsx(regions_file))
cat('Loading coverage matrix..\n')
covmat <- preloadCovMatrix(coverage_matrix)
```

```{r}
cnvCount=length(CNVs$position)
for (i in 1:cnvCount ) {
  rchr=CNVs[i,'chrom']
  rstart=CNVs[i,'start']
  rend=CNVs[i,'stop']
  outpdf=paste0(outdir, '/', sprintf('%02d',i),'_chr', rchr,"_",rstart,"_",rend, out_suffix, '.pdf')
  cat('\n>>>> Processing region ', i,' of ',cnvCount, ' : chr',rchr, ':', rstart, '-', rend, '\n')
  CNViewMod(chr=rchr,start=rstart, end = rend,
         #region to be plotted
         sampleID=common_brains,   #Character vector of IDs of samples to plot
         covmatrix=covmat,
         compression="optimize",   #compression factor for rebinning, if desired
         highlight=NA,             #list of coordinate pairs; intervals to highlight; defaults to query interval; NULL disables
         highlightcol="gold",      #vector of colors to shade each highlighted interval
         window=NA,                #distance to append to both sides of input interval for viewing; NA = 61.8% on either side
         yscale="optimize",        #vector of values to be represented on y axis
         normDist=10000000,         #distance outside region to normalize (both sides). Must either be int or "genome"
         subsample=200,            #will only load this many samples into memory; useful to reduce runtime & memory reqs for very large cohorts
         UCSCtracks=c("Gene",      #append UCSC sequence context information; choose either NULL
                      "SegDup",    #or up to three among "Gene", "Gap", "RepMask", "blacklist", and "SegDup"
                      "Gap"),
         genesymbols=TRUE,         #print gene symbols below UCSC gene body annotations
         probs=TRUE,               #option to add CNV probabilities below each highlighted interval
         gcex=1,                   #global scaling for all fonts
         title=NULL,               #option to add custom title. Overrides default
         panelnames=NA,            #optional vector for custom names printed above each plot; only works for multi-sample plots
         legend=T,                 #logical option to plot legend
         output=outpdf,              #path to output as pdf. If NULL, will plot to active device
         plot=TRUE,                #logical option to disable plot step; mandatory for output!=NULL
         returnData=FALSE,         #logical option to return df of all normalized coverage values
         quiet=FALSE)
}

```

Combine pdfs:
```{r}
final_pdf=paste0(outdir, '/', gsub('^_','', out_suffix),'.joined.pdf')
cat ("Combining PDFs...")
fpdf <- pdf_combine(list.files(path = outdir, full.names = TRUE, pattern=paste0('*',out_suffix,'.pdf')), 
            output =  final_pdf)
cat(" Done.\nFinal PDF is in:", fpdf ,"\n")
```
